SIMD-0167: Loader-v4 (#167)
solana-foundation/solana-improvement-documents: proposals/0167-loader-v4.md

    - Enum variant `1u64`: Deployed, program is ready to be executed
    - Enum variant `2u64`: Finalized, same as `Deployed`, but can not be
    modified anymore
- Body:
  - `[u8]` The programs executable file

Verification the program account checks in the following order that:

- the owner of the program account is loader-v4,
otherwise throw `InvalidAccountOwner`
- the program account is at least as long enough for the header,
otherwise throw `AccountDataTooSmall`
- the program account is writable, otherwise throw `InvalidArgument`
- the provided authority (instruction account at index 1) signed,
otherwise throw `MissingRequiredSignature`
- the authority stored in the program account is the one provided,
otherwise throw `IncorrectAuthority`
- the status stored in the program account is not finalized,
otherwise thorw `Immutable`

### Execution / Invocation

Invoking programs owned by loader-v4 checks in the following order that:

- the owner of the program account is loader-v4,
otherwise throw `UnsupportedProgramId`
- the program account is at least as long enough for the header,
otherwise throw `AccountDataTooSmall`

> - the status stored in the program account is not retracted,

Author: The check asserts that the status is **not** `retracted`, thus only `retracted` throws the error.

  would load exactly one account, no need to load the proxy account to get to
  the actual program data (which is not listed in the transaction accounts).
  - allows the removal of the write lock demotion exception if loader-v3 is
  present in a transaction.
  - corrects the miscounting of the proxy account size towards the total
  transaction account loading limit.

## Security Considerations

None.

## Backwards Compatibility

This proposal does not break any existing programs. However, dapp developers
might want to profit from the new program mangement instructions without
influencing their users work flows. To do so they would need a way to turn the
program accounts of loader-v3 to program accounts of loader-v4, changing the
account owner but keeping the program address. A potential issue is that the
programdata header of loader-v3 is only 45 bytes long while loader-v4 takes 48
bytes. An automatic mechanism in the program runtime (triggered by feature
activation) could then perform the following steps per program:

- loader-v3 clears the program proxy account (setting its size to zero)
- loader-v3 transfers all funds from the programdata to the proxy account

> - loader-v3 gifts the program proxy account to loader-v4

Reviewer: How would the gifting work? Do we need to add a new instruction to the v3 loader? Will this upgrade mechanism be specified in a future SIMD?

Author: I see three options here:

1. A migrate instruction for loader-v3 which the dApp devs trigger themselves
2. A migrate instruction for loader-v3 which we the protocol engineers trigger
3. A feature gate which scans the accounts DB and migrates all loader-v3 programs

I would prefer the second option as that allows us to keep the gathering of the loader-v3 program list and then slowly working through it off-chain and still makes sure that all loader-v3 programs get migrated.

Author: Added a more detailed description of how this would work:
https://github.com/solana-foundation/solana-improvement-documents/pull/167/commits/4f5de707a5d5d2a91dac4efe650c0073d205a4cb

Consequentially, this setup of the program account being a proxy account,
containing the address of the actual program data account, is no longer
necessary and should be removed. Likewise the distinction of executable program
accounts and non-executable buffer accounts is also no longer necessary and
should be removed as well.

In loader-v3 every instruction which modified the program data had to re-verify
the ELF in the end. Instead we are now aiming for a more modular workflow which
explicates these steps in a Retract-Modify-Redeploy sequence of instructions
that allows multiple modifications to share one verification of the ELF in the
end. This sequence can be a single transaction or split across multiple
transactions.

Another issue with loader-v3 is that the executable file stored in the
programdata account is misaligned relative to the beginning of the account.
This currently requires a copy in the ELF loader to re-align the program.

Additionally, there currently is no complete specification of the loaders
program management instructions. This proposal would thus fill that gap once
loader-v4 goes into production.

See impact for further motivation.

> ## Alternatives Considered

Reviewer 2: One thing I haven't been able to understand is why we added a new version of the loader instead of a new version of the program account state. Can you write up why we couldn't add a `UpgradeableLoaderState::ProgramV2` which doesn't have a programdata and is properly aligned? Seems like that would have been the simplest fix.

Author: Added

Reviewer 3: Is it possible to add more details about the "special casing" that adding a new status makes it harder than a new loader? The existence of the `UpgradeableLoaderState` should have made this easier. Will loader v4 need "special casing" as well?

A delay-visibility-free redeployment could be achieved by keeping the swap
program around until the end of the slot. This would however mean that two
accounts per program must be loaded until the dapp developer reclaims the
second one. That would defy the purpose of this proposal which is to get rid
of the proxy account.

## New Terminology

None.

> ## Detailed Design

Reviewer 2: I don't see any mention about loader-v4 instructions being prohibited from being invoked via cpi so this would be the first time we allow setting the `is_executable` flag on an account in a CPI program execution. If an sbf program invoked a cpi deploy, it wouldn't be able to see that flag changing because we currently don't update the `is_executable` flag for the sbf caller program. We should either prohibit any instructions that change the `is_executable` flag from being cpi's or we should update the SVM to copy the value back into the caller's memory.

  - Check that the end offset (sum of offset and length of the chunk) does
  not exceed the maximum (program account length minus the header size),
  otherwise throw `AccountDataTooSmall`
  - Copy the chunk into the program account at the offset shifted by the
  header size

#### Copy

- Instruction accounts:
  - `[writable]` The program account to copy to.
  - `[signer]` The authority of the program.
  - `[]` The program(data) account to copy from.
- Instruction data:
  - Enum variant `1u32`
  - `u32` Byte offset at which to write
  - `u32` Byte offset at which to read
  - `u32` Length of the chunk to copy in bytes
- Behavior:
  - Check there are at least three instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Check that program account and source account do not alias,
  otherwise throw `AccountBorrowFailed`
  - Verify the program account
  - Check the status stored in the program account is retracted,
  otherwise throw `InvalidArgument`
  - Check that the source account is owned by loader v1, v2, v3 or v4,
  otherwise throw `InvalidArgument`
  - and look-up the source header size:
    - loader-v1: 0 bytes
    - loader-v2: 0 bytes
    - loader-v3: 45 bytes

>     - loader-v4: 48 bytes

Author: Yes, good point. I simply did not consider that anybody would like to use loader-v3 buffers to deploy loader-v4 programs as that is not necessary for the migration.

  - Check that the source end offset (sum of source offset and length) does
  not exceed the maximum (source account length minus the source header size),
  otherwise throw `AccountDataTooSmall`
  - Check that the destination end offset (sum of destination offset and
  length) does not exceed the maximum (program account length minus the loader-v4
  header size), otherwise throw `AccountDataTooSmall`
  - Copy the chunk between the program accounts at the offsets, each shifted by
  the header size of their loader (account owner) respectively

#### SetProgramLength

- Instruction accounts:
  - `[writable]` The program account to change the size of.
  - `[signer]` The authority of the program.
  - `[writable]` Optional, the recipient account.
- Instruction data:
  - Enum variant `2u32`
  - `u32` The new size after the operation.
- Behavior:
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - If this is an initialization (program account length is too short to

>   contain the header):

Author: We can add an explicit `Uninitialized` state to the enum.

    - the owner of the program account is loader-v4,
    otherwise throw `InvalidAccountOwner`
    - the program account is writable, otherwise throw `InvalidArgument`
    - the provided authority (instruction account at index 1) signed,
    otherwise throw `MissingRequiredSignature`
  - If this is not an initialization:
    - Verify the program account
    - Check that the status stored in the program account is retracted,
    otherwise throw `InvalidArgument`
  - Check that there are enough funds in the program account for rent
  exemption, otherwise throw `InsufficientFunds`
  - If there are more than enough funds:
    - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
    - Check that the recipient account (instruction account at index 2) is
    writable, otherwise throw `InvalidArgument`
    - If a recipient account was provided that is not the program account:
      - Transfer the surplus from the program account to the recipient account
    - otherwise, if the requested new size is zero throw `InvalidArgument`

>   - If the requested new size is zero:

Reviewer 2: Can you specify more explicitly how a program account should be deleted?
1. reduce lamports to zero
2. set data length to zero
3. possibly reset `is_executable` flag

Reviewer 2: > The account is deleted implicitly at the end of the transaction 

It might not be implicitly deleted at the end of the transaction if its lamport balance is topped up by a later instruction in the transaction. The loader-v4 program cannot guarantee that the program account will be deleted. All it can guarantee is the state of certain account fields like the data, lamports, owner, and executable flag. So I think this SIMD should be explicit about what exactly the loader-v4 program will be changing about the account rather than stating a (potentially false) assumption about the account getting deleted later.

Reviewer 2: That would be nice and also please list the account modifications to make the spec explicit (lamports, data length, etc)

>     - Delete the entire program account, including the header

Reviewer 2: Even if the program account is deleted, its `is_executable` flag will still be set to `true` and another instruction could transfer lamports to the account to keep it alive. This would create a weird edge case where a loader-v4 owned account would have its `is_executable` flag set to `true` despite not having a program state.

Author: We don't assign the account to the system program, so the actual deletion only happens implicitly at the end of the transaction. In between, no other instruction in the transaction which is not running loader-v4 can do anything with the account.

Reviewer 2: Yeah, my point is that a _later_ instruction in the transaction could top up the balance of the closed program account such that it doesn't get deleted at the end of the transaction.

Author: Then it would stay a loader-v4 program. So the only thing you could get is an uninitialized loader-v4 program which has the executable flag set. However, you can already have a retracted loader-v4 program which has the executable flag the "official" way, and that is quite similar.

Reviewer 2: Yes, exactly. This is what I said is the issue:
> This would create a weird edge case where a loader-v4 owned account would have its is_executable flag set to true despite not having a program state

Author: The plan is to emulate the `is_executable` flag in the VM using only the owner field, regardless of the program state. So this is already consistent with that goal.

Reviewer 2: Sorry I don't really understand your response

Reviewer 2: As long as `is_executable` is set and readable in the VM, I think we should go with because it's the most "correct" and not as complex as the verified option you mentioned.
> Make it true depending on the loader state: This is more complex and depends on the loaders.

If we go ahead with further cleanup of the `is_executable` flag in a new SIMD, I think we should do that first and then this SIMD doesn't need to worry about the flag at all.

Author: Then this SIMD would have to specify that:

- The flag is set when the program is deployed
- The flag is reset when the program is retracted

Reviewer 2: Hmm maybe? I don't really understand why loader-v4 needs the ability to "retract" programs in addition to closing them.

Author: The retract state is what tells the program runtime that a program needs to be re-verified before it can be deployed. That way we can allow multiple modifications before one shared verification. Also, it allows a modular composition of these modifications instead of requiring a special instruction for each use-case / combination of those.

of the proxy account.

### Adding a loader-v3 status

Instead of adding a loader-v3 status a new loader was added for the following
reasons:

- There is a lot of special casing throughout the validator for loader-v3,
which does not apply to loader-v4.
- The existing instruction interface of loader-v3 (e.g. the distinction
of program account and programdata account) are incompatible with loader-v4.
- Adding a new loader-v3 status would create new unforseen interactions with
the existing status enum and instructions.
- Modifying the existing loader-v3 implementation risks breaking it, while
a independent new loader does not.
- Loader-v3 is undocumented and unspecified. Starting fresh allows to have
a complete implementation, specification and documentation.

## New Terminology

None.

## Detailed Design

The feature gate must:

- enable loader-v4 `LoaderV411111111111111111111111111111111111` program
management and execution.
- enable the loader-v3 `BPFLoaderUpgradeab1e11111111111111111111111`
instruction `UpgradeableLoaderInstruction::Migrate` (see SIMD-0315).

### Owned Program Accounts

Accounts of programs owned by loader-v4 must have the following layout:

> - Header (which is 48 bytes long):

Author: I mean, yes we could, but what use-case would it actually aid?
Let's say you can change the account size beyond a program independently of redeployment and re-verification.
But subsequently, you would still have to redeploy to upload a longer program, and that would still require re-verification in the end.

Also I think we addressed the Retract-Modify-Redeploy pattern that @buffalojoec was concerned about with the split of the `NeverBeenDeployed` and `Retracted` status.

> - Header (which is 48 bytes long):

Reviewer 4: Overall I think it's a better flow if only those instructions that modify an ELF require verification. Nobody wants to have to pay CUs for two verifications to do one deployment. Furthermore, couldn't you just deploy a super big program and keep extending it by one byte to grief the validator?

> - Header (which is 48 bytes long):

Author: > have to pay CUs for two verifications to do one deployment

That is exactly what the Retract-Modify-Redeploy pattern avoids.

> Furthermore, couldn't you just deploy a super big program and keep extending it by one byte

No because verification only occurs at `Deploy` and that has a cooldown so it can only be done once per slot.

> - Header (which is 48 bytes long):

Reviewer 2: If we store the length, then we don't need to do the complicated Retract-Modify-Redeploy pattern at all. We just have a single "Redeploy" which verifies the source buffer, sets the new length, and deploys.

> - Header (which is 48 bytes long):

Reviewer 2: I get your argument about monolithic instructions being hard to maintain and audit but there is also inherent complexity in a modular approach because now auditors and maintainers have to consider many different combinations and orderings of instructions that could have unintended side effects.

With a modular approach we are also doing a lot of the same checks in each instruction that could be done in one instruction. Also consider that the cost of using CPI's and PDA's across multiple instructions for a single logical operation is more expensive (this would be the case for a multisig owned program).

And beyond perf / cu's, I don't think we should push more complexity to developers by forcing them to know that in order to deploy, you must first invoke a retract instruction, for example. They are already used to being able to invoke an upgrade instruction with a buffer, I think we should at least make it possible to do that with a single instruction.

> - Header (which is 48 bytes long):

Author: Yes, it does cost more CUs and requires more CPI to have it split in smaller instructions. I can add that to the considerations section. Yet, program management instructions are already rare and expensive, even rate limited. Thus, I don't see the cost of the actual instructions as the main priority here, especially since the required rent exemption is the largest part anyway.

The modular approach can also maintain invariants which prune a whole lot of the combinatoric space and make it easy to reason about (e.g. the added program status field).

About usability, I think its a fair point, yet I don't see that much of a difference for the user: The CLI can hide most of it away and for the few DAO like and self upgrading programs it is the difference between figuring out all the parameters of one big instruction vs figuring out the order of more smaller instructions. Also, we do include a list of common work flows in the SIMD so dApp devs don't even have to figure this out on their own.

> - Header (which is 48 bytes long):

Reviewer 2: Your arguments make sense to me, thanks for the response

  version of the program.
- Body:
  - `[u8]` The programs executable file

Verification the program account checks in the following order that:

- the owner of the program account is loader-v4,
otherwise throw `InvalidAccountOwner`
- the program account is at least as long enough for the header,
otherwise throw `AccountDataTooSmall`
- the program account is writable, otherwise throw `InvalidArgument`
- the provided authority (instruction account at index 1) signed,
otherwise throw `MissingRequiredSignature`
- the authority stored in the program account is the one provided,
otherwise throw `IncorrectAuthority`
- the status stored in the program account is not finalized,
otherwise throw `Immutable`

### Execution / Invocation

Invoking programs owned by loader-v4 checks in the following order that:

- the owner of the program account is loader-v4
- the program account is at least as long enough for the header
- the status stored in the program account is deployed or finalized
- the program account was not deployed within the current slot (delay
visibility)
- the executable file stored in the program account passes executable

> verification

Author: No, the feature set can change and thus programs can become valid or invalid after deployment.

  executable verification
  - Change the slot in the program account to the current slot
  - Change the status stored in the program account to deployed
  - Set the `is_executable` flag to `true`

#### Retract

- Instruction accounts:
  - `[writable]` The program account to retract.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `4u32`
- Behavior:
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the slot stored in the program account is not the current
  (deployment cooldown), otherwise throw `InvalidArgument`
  - Check that the status stored in the program account is deployed,
    otherwise throw `InvalidArgument`
  - Note: The slot is **not** set to the current slot to allow a
  retract-modify-redeploy-sequence within the same slot or even within the
  same transaction.
  - Change the status stored in the program account to retracted
  - Set the `is_executable` flag to `false`

#### TransferAuthority

- Instruction accounts:
  - `[writable]` The program account to change the authority of.
  - `[signer]` The current authority of the program.

>   - `[signer]` The new authority of the program.

Reviewer 5: I assume the reason to remove the _unchecked_ version of this instruction is due to concerns of losing the authority. Assuming this is the case, could we have an alternative 2-step process (TransferAuthority + AcceptTransferAuthority)? 

With the ideas around removing nonces, requiring both signatures is going to further increase developer frustrations around program upgrades.

Author: How would you implement that? It seems it would require a transitory state (`InAuthorityTransfer`) which would require two authority fields in the header in case the new one does not accept and you need to revert.

- the executable file stored in the program account passes executable
verification

failing any of the above checks must throw `UnsupportedProgramId`.

### Program Management Instructions

All program management instructions must cost 2000 CUs.

#### Write

- Instruction accounts:
  - `[writable]` The program account to write to.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `0u32`
  - `u32` Byte offset at which to write the given bytes
  - `[u8]` Chunk of the programs executable file
- Behavior:
  - Check there are at least two instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check the status stored in the program account is never-been-deployed or
  retracted,
  otherwise throw `InvalidArgument`
  - Check that the end offset (sum of offset and length of the chunk) does
  not exceed the maximum (program account length minus the header size),
  otherwise throw `AccountDataTooSmall`
  - Copy the chunk into the program account at the offset shifted by the
  header size

#### Copy

- Instruction accounts:
  - `[writable]` The program account to copy to.
  - `[signer]` The authority of the program.

>   - `[]` The program(data) account to copy from.

Author: No this is correct, in case of loader-v3 you copy from a programdata account, not a program account.

>   - `[]` The program(data) account to copy from.

Reviewer 4: The point of this spec was to avoid the v3 spec. I don't really see a reason to reference anything about any other loader specs in this proposal.

>   - `[]` The program(data) account to copy from.

Author: There is no loader-v3 spec ;)

I can take it out.

- Instruction data:
  - Enum variant `1u32`
  - `u32` Byte offset at which to write
  - `u32` Byte offset at which to read
  - `u32` Length of the chunk to copy in bytes
- Behavior:
  - Check there are at least three instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Check that program account and source account do not alias,

>   otherwise throw `AccountBorrowFailed`

Author: This is what is generally emitted by the program runtime whenever borrows do alias. Thus it is consistent.

>   otherwise throw `AccountBorrowFailed`

Author: https://github.com/anza-xyz/agave/blob/8854e65d81b8d74c6e55e292b93e3a66e72a4c81/transaction-context/src/lib.rs#L793
This is the standard error code for all aliasing.

    - the owner of the program account is loader-v4,
    otherwise throw `InvalidAccountOwner`
    - the program account is writable, otherwise throw `InvalidArgument`
    - the provided authority (instruction account at index 1) signed,
    otherwise throw `MissingRequiredSignature`
  - If this is not an initialization:
    - Verify the program account
    - Check that the status stored in the program account is
    never-been-deployed or retracted,
    otherwise throw `InvalidArgument`
  - Check that there are enough funds in the program account for rent
  exemption of the new length, otherwise throw `InsufficientFunds`
  - Set the length of the program account to the requested new size plus
  the header size
  - In case that this is an initialization, also initialize the header:
    - Set the slot to zero, **not** the current slot
    - Set the authority address (from the instruction account at index 1)
    - Set the status to never-been-deployed

#### WithdrawLamports

- Instruction accounts:
  - `[writable]` The program account to withdraw from.
  - `[signer]` The authority of the program.
  - `[writable]` The recipient account.
- Instruction data:
  - Enum variant `3u32`

>   - `u64` The amount in lamports.

Reviewer 4: What do you think about just skipping this argument and always withdrawing everything over rent-exemption? Seems strange for someone to only withdraw a little bit of the excess.

- Behavior:
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Check that program account and recipient account do not alias,

>     otherwise throw `AccountBorrowFailed`

Reviewer 4: ```suggestion
    otherwise throw `InvalidArgument`
```

  - Verify the program account
  - Check that there would be enough funds in the program account remaining for
  rent exemption after the withdrawl, otherwise throw `InsufficientFunds`
  - If all lamports are being withdrawn:
    - Check that the status stored in the program account is
    never-been-deployed or retracted,
    otherwise throw `InvalidArgument`

>     - Set the length of the program account to 0 (removing the header too)

Reviewer 4: Ah, I see why you're requiring the argument for number of lamports. This feels like it belongs in a dedicated `Erase` instruction. The `Erase` instruction can resize the account to 0 and withdraw all lamports.

  - Transfer the amount from the program account to the recipient account

#### Deploy

- Instruction accounts:
  - `[writable]` The program account to deploy.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `4u32`
- Behavior:
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the slot stored in the program account is not the current
  (deployment cooldown), otherwise throw `InvalidArgument`
  - Note: The cooldown enforces that each pair of an address and a slot can
  uniquely identify a deployment of a program, which simplifies caching logic.

>   - Check that the status stored in the program account is never-been-deployed
>   or retracted

Reviewer 4: ```suggestion
  - Check that the status stored in the program account is
    `NeverBeenDeployed` (`1u64`) or `Retracted` (`2u64`),
```

    otherwise throw `InvalidArgument`
  - Check that the executable file stored in the program account passes
  executable verification

>   - Change the slot in the program account to the current slot

Reviewer 4: ```suggestion
  - Change the slot in the program account to the current slot on the Clock sysvar
```

>   - Change the status stored in the program account to deployed

Reviewer 4: ```suggestion
  - Change the status stored in the program account to `Deployed` (`3u64`)
```

  - Set the `is_executable` flag to `true`

#### Retract

- Instruction accounts:
  - `[writable]` The program account to retract.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `5u32`
- Behavior:
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the slot stored in the program account is not the current
  (deployment cooldown), otherwise throw `InvalidArgument`
  - Check that the status stored in the program account is deployed,
    otherwise throw `InvalidArgument`
  - Note: The slot is **not** set to the current slot to allow a
  retract-modify-redeploy-sequence within the same slot or even within the
  same transaction.

>   - Change the status stored in the program account to retracted

Reviewer 4: ```suggestion
  - Change the status stored in the program account to `Retracted` (`2u64`)
```

  - Set the `is_executable` flag to `false`

> #### TransferAuthority

Reviewer 4: A well-known standard is to call this kind of operation "SetAuthority". What do you think?

```suggestion
#### SetAuthority
```

- Instruction accounts:
  - `[writable]` The program account to change the authority of.
  - `[signer]` The current authority of the program.

>   - `[(signer)]` The new authority of the program.

Reviewer 4: ```suggestion
  - `[optional(signer)]` The new authority of the program.
```

- Instruction data:
  - Enum variant `6u32`
- Behavior:
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the new authority (instruction account at index 2)
  is either the system program or has signed,
  otherwise throw `MissingRequiredSignature`
  - Check that the authority stored in the program account is different
  from the one provided, otherwise throw `InvalidArgument`
  - Copy the new authority address into the program account
  - If the the new authority is the system program:
    - Change the status stored in the program account to finalized

### Workflows

#### Inital deployment

- Assign account to loader-v4

> - SetProgramLength to ELF size

Reviewer 4: You can also do `System::CreateAccount` here!

### Program Management Instructions

All program management instructions must cost 2000 CUs.

#### Write

- Instruction accounts:
  - `[writable]` The program account to write to.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `0u32`
  - `u32` Byte offset at which to write the given bytes
  - `[u8]` Chunk of the programs executable file
- Behavior:
  - Check there are at least two instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check the status stored in the program account is retracted,
  otherwise throw `InvalidArgument`
  - Check that the end offset (sum of offset and length of the chunk) does
  not exceed the maximum (program account length minus the header size),
  otherwise throw `AccountDataTooSmall`
  - Copy the chunk into the program account at the offset shifted by the
  header size

#### Copy

- Instruction accounts:
  - `[writable]` The program account to copy to.
  - `[signer]` The authority of the program.
  - `[]` The program(data) account to copy from.
- Instruction data:
  - Enum variant `1u32`
  - `u32` Byte offset at which to write
  - `u32` Byte offset at which to read

>   - `u32` Length of the chunk to copy in bytes

Author: I think there is no consistent style. [memcpy](https://www.man7.org/linux/man-pages/man3/memcpy.3.html) for example is dest then src.

  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - If this is an initialization (program account length is too short to
  contain the header):
    - the owner of the program account is loader-v4,
    otherwise throw `InvalidAccountOwner`
    - the program account is writable, otherwise throw `InvalidArgument`
    - the provided authority (instruction account at index 1) signed,
    otherwise throw `MissingRequiredSignature`
  - If this is not an initialization:
    - Verify the program account
    - Check that the status stored in the program account is retracted,
    otherwise throw `InvalidArgument`
  - Check that there are enough funds in the program account for rent
  exemption, otherwise throw `InsufficientFunds`
  - If there are more than enough funds:
    - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
    - Check that the recipient account (instruction account at index 2) is
    writable, otherwise throw `InvalidArgument`
    - If a recipient account was provided that is not the program account:
      - Transfer the surplus from the program account to the recipient account

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Author: Typically, you would only free up funds to be withdrawn when you decrease the size and thus the rent exemption cost.

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Author: What would you gain from having it be a separate instruction? Like what use case does it fulfill? Currently the instruction does not mandate any length change, so you can already use it to withdraw only without changing your program.

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Reviewer 6: +1 splitting out lamports handling. one should not have it issue an instruction about resizing the program to recover mistransferred lamports. they're simple enough to compose at the tx level for the common case

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Author: While I agree on a conceptual level, I don't think it matters that much one way or the other. You have to keep in mind that this entire thing was already implemented roughly a year ago, is tested and reviewed, coordinated across validator implementations and active on testnet. So if you really think it is that important, we can make another SIMD to change this. But there is no point in halting the rollout now. Especially since before (loader-v3) you couldn't get the funds back at all, so the main pain point is already addressed.

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Reviewer 4: I feel like this SIMD should have been merged before the rollout, though. Seeing it open and not yet approved on the Anza side would make me think it's still under review.

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Author: Yes, that is true, formalism was not adhered to correctly regarding the merging. Yet, I simply can't approve my own PR.

>     - otherwise, if the requested new size is zero throw `InvalidArgument`

Reviewer 4: Considering offline discussions and the status of the testnet & devnet features, I think we can still split this out.

  - If the requested new size is zero:
    - Delete the entire program account, including the header
  - If the requested new size is greater than zero:
    - Set the length of the program account to the requested new size plus
    the header size
    - In case that this is an initialization, also initialize the header:

>       - Set the `is_executable` flag to `true`

Author: SIMD-0162 only removes the checks of the `is_executable` flag in the runtime. It does not remove the `is_executable` flag entirely. E.g. it is still exposed to the VM. Cleaning that up will require another SIMD.

>       - Set the `is_executable` flag to `true`

Author: That is because it is in line with what the older loaders did. That way there is less inconsistency when programs read that flag after they migrate.

>       - Set the `is_executable` flag to `true`

Reviewer 4: The entire motivation behind loader v4 is to break away from the concepts that the other loaders have enshrined, so why not also break away from this one? It's going to happen eventually.

>       - Set the `is_executable` flag to `true`

Author: The removal of the `is_executable` flag will have its own SIMD and the flag will be emulated by checking the account owner instead. The outcome will be that it stays `true` (from the pov inside the VM) for all loaders (including loader-v4), but the program runtime won't have that field anymore. Then in ABIv2 we will also remove that field for the programs inside the VM.

Other way around: If we delete this line from the SIMD, it means we implicitly set it to `false`. Then in the removal of the `is_executable` flag, it will be emulated and turn back to `true`. So, you would transition through the two SIMDs: `true`, `false`, `true`. No point in doing that.

>       - Set the `is_executable` flag to `true`

Reviewer 4: > Then in the removal of the `is_executable` flag, it will be emulated and turn back to `true`.

Maybe I'm missing something with the other `is_executable` SIMDs, but when you say "emulating", aren't you simply saying it's just an ownership check evaluation? In other words, the field isn't emulated, but _ignored_ (evaluates as `true` as a result of the ownership check). If that's the case, it shouldn't matter what the field's value is anymore, and this SIMD shouldn't have it in the spec if it's deprecated on the protocol.

>       - Set the `is_executable` flag to `true`

Author: There exists an internal `is_executable` inside the program runtime, that is now being ignored. And there exists an external `is_executable` which is visible to the programs inside the VM, which still needs to have a value. Currently, it is the value stored in the accounts DB and later it will be emulated (by ownership check).

  - Note: The slot is **not** set to the current slot to allow a
  retract-modify-redeploy-sequence within the same slot or even within the
  same transaction.
  - Change the status stored in the program account to retracted

#### TransferAuthority

- Instruction accounts:
  - `[writable]` The program account to change the authority of.
  - `[signer]` The current authority of the program.
  - `[signer]` The new authority of the program.
- Instruction data:
  - Enum variant `5u32`
- Behavior:
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the new authority (instruction account at index 2)
  signed as well, otherwise throw `MissingRequiredSignature`
  - Check that the authority stored in the program account is different
  from the one provided, otherwise throw `InvalidArgument`
  - Copy the new authority address into the program account

#### Finalize

- Instruction accounts:
  - `[writable]` The program account to change the authority of.
  - `[signer]` The current authority of the program.
  - `[]` Optional, the reserved address for the next version of the program.

> - Instruction data:

Reviewer 4: If any ecosystem projects weigh in on this design and express support, I say we do it. But we shouldn't require the keypair to be held from the time the current version is frozen (`Finalized`), in my opinion.

> - Instruction data:

Author: We don't know if this will be used at all or if anybody cares. Right now it incurs no costs on anybody and is purely opt-in. If the community engages and starts using this, then we might get a better idea of how to adjust it.

> - Instruction data:

Reviewer 4: > We don't know if this will be used at all or if anybody cares.

OK this was my reason for asking in the first place, though. I don't think we should add features that we aren't sure people need.

  - Enum variant `6u32`
- Behavior:
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the status stored in the program account is deployed,
    otherwise throw `InvalidArgument`
  - for the program account of the next version
  (instruction account at index 2) check that:
    - the owner of the program account is loader-v4,
    otherwise throw `InvalidAccountOwner`
    - the program account is at least as long enough for the header,
    otherwise throw `AccountDataTooSmall`
    - the authority stored in the program account is the one provided,
    otherwise throw `IncorrectAuthority`
    - the status stored in the program account is not finalized,
    otherwise throw `Immutable`
  - Copy the address of the next version into the next version field stored in
  the previous versions program account
  - Change the status stored in the program account to finalized

> ### Loader-v3 Instruction: Migrate

Author: It is not just any loader-v3 instruction. It is specifically related, because it is the loader-v3 -> loader-v4 migration instruction. It is also behind the same feature gate, that is why I put it in the same SIMD.

feature: 2aQJYqER2aKyb3cZw22v4SL2xMX7vwXBRWfvS4pTrtED
---

## Summary

A new upgradeable loader which only requires a single account per program.

## Motivation

Loader-v3, which is currently the only deployable loader, requires two accounts
per program. And has three account types: Program, program-data and buffer.
This was a workaround to circumvent the finality of the `is_executable` flag,
which will be ignored by the program runtime from SIMD-0162 onwards.
Consequentially, this setup of the program account being a proxy account,
containing the address of the actual program data account, is no longer
necessary and should be removed. Likewise the distinction of executable program
accounts and non-executable buffer accounts is also no longer necessary and
should be removed as well.

In loader-v3 every instruction which modified the program data had to re-verify
the ELF in the end. Instead we are now aiming for a more modular workflow which
explicates these steps in a Retract-Modify-Redeploy sequence of instructions
that allows multiple modifications to share one verification of the ELF in the
end. This sequence can be a single transaction or split across multiple

> transactions.

Reviewer 2: Was this discussion ever resolved?

    - CPI loader-v4 `SetProgramLength` the program account to the size of the
    program data account minus the loader-v3 header size and use the migration
    authority.
    - CPI loader-v4 `Copy` the program data account into the program account
    - CPI loader-v4 `Deploy` the program account
    - If the program data account was finalized (upgrade authority is `None`):
      - CPI loader-v4 `Finalize` without a next version forwarding
    - otherwise, if the program data account was not finalized and the
    migration authority (as opposed to the upgrade authority) was provided:
      - CPI loader-v4 `TransferAuthority` to the upgrade authority
  - Clear the program data account (setting its size to zero)
  - Assign ownership of the program data account to the system program

## Impact

- This proposal covers all the use cases loader-v3 had but in a cleaner way and
comes with a specification.
- loader-v3 had a separate account type for buffers and extra commands for
these buffer accounts, in loader-v4 program accounts can act as buffers, there
is no more distinction.
- loader-v3 deployments always needed a buffer, in loader-v4 it is optional,

> one can upload a redeployment into the program account directly.

Author: First of all, you still have to send a `Retract` instruction first, before you can start modifying the program. In that sense the `Retracted` state is like the loader-v3 buffer and the `Deployed` state is like the loader-v3 program.

If accidental change of a deployed program is a concern, we can add a warning and confirmation mechanism to the CLI.

> one can upload a redeployment into the program account directly.

Reviewer 4: > `Retracted` state is like the loader-v3 buffer and the `Deployed` state is like the loader-v3 program

This isn't true, though. `Retracted` means "disabled", and that leaves your program in a state where, should anything go wrong with any of the writes, it's inactive until you solve the problem.

Buffers, on the other hand, don't change anything about the program's liveness.

redeployment.
- loader-v3 always required the entire program to be uploaded for a
redeployment, loader-v4 supports partial uploads for patching chunks of the
program.
- loader-v3 ELFs were misaligned, loader-v4 properly aligns the executable
file relative to the beginning of the account.
- loader-v4 allows finalized programs to mark which other program supersedes
them which can then be offered as an option in frontends. This provides a
more secure alternative to redeployment / upgrading of programs at the same
address. The keypair for the next version linked during finalization should be
generated beforehand.
- An option to migrate programs from loader-v3 to loader-v4 without changing
their program address will be available via a new loader-v3 instruction. This
will count as a redeployment and thus render the program unavailable for the
rest of the slot (delay visibility).

Once new programs can not be deployed on loader-v3 anymore, the list of all
loader-v3 programs becomes fixed and can be extracted from a snapshot. Using
the added loader-v3 migration instruction and the global migration authority,
the core protocol developers will then migrate all loader-v3 programs to

> loader-v4 programs, which once completed:

Author: https://github.com/solana-foundation/solana-improvement-documents/pull/315

program around until the end of the slot. This would however mean that two
accounts per program must be loaded until the dapp developer reclaims the
second one. That would defy the purpose of this proposal which is to get rid
of the proxy account.

## New Terminology

None.

## Detailed Design

The feature gate `8Cb77yHjPWe9wuWUfXeh6iszFGCDGNCoFk3tprViYHNm` must:

- enable loader-v4 `LoaderV411111111111111111111111111111111111` program
management and execution.
- enable the loader-v3 `BPFLoaderUpgradeab1e11111111111111111111111`
instruction `UpgradeableLoaderInstruction::Migrate`.

An additional feature gate `EmhbpdVtZ2hWRGFWBDjn2i3SJD8Z36z4mpMcZJEnebnP`
must disable new deployments on loader-v3,
throwing `InvalidIstructionData` if `DeployWithMaxDataLen` is called.

### Owned Program Accounts

Accounts of programs owned by loader-v4 must have the following layout:

- Header (which is 48 bytes long):
  - `u64` Slot in which the program was last deployed, retracted or
  initialized.
  - `[u8; 32]` Authority address which can send program management
  instructions. Or if the status is finalized, then the address of the next

>   version of the program.

Reviewer 7: This seems to mean that a program is either upgradable or finalized with the new version already specified.
There is no way to finalize a program separately from setting the new version address.

Obviously, there is more than one way how people can manage their programs.
And maybe this is something that has been already discussed.
But it seems that a number of projects use non-upgradable programs, yet they do deploy new versions later.

They would not be able to specify a new version address, if they deploy and finalize their programs before the new version is available.

Author: That is why the reserve instruction says "the reserved address for the next version of the program". Meaning one has to setup an empty program first and then link to that during finalization. Then later deploy an actual program there.

Modifying the next version link after finalization would break the immutability of finalized program accounts.

Reviewer 7: Indeed, they can create the forwarding keypair beforehand.  Have not though about it.
Maybe add an informal note to clarify that this is a solution people can use here?

It still seems a bit suboptimal that you can not tell if a new version exists by just looking at the current version.
For immutable programs that have forwarding addresses, you effectively always check the forwarding address content, and it is expected that the target may not exist yet.

It would be a more complex version, but I was thinking that an alternative approach would be to add two separate `Finalized` states.
Say, `FinalizedNoForwarding` and `FinalizedWithForwarding`.
In the `FinalizedNoForwarding` state, the data is immutable, but the second field is still the authority key.
While in the `FinalizedWithForwarding` state, the second field is the forwarding address.

`FinalizedNoForwarding` does not allow for the data to be modified.
But if the authority key is non-zero, it allows an update that changes the state to `FinalizedWithForwarding` and overwrites the authority key with a proper forwarding address.

    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the slot stored in the program account is not the current
  (deployment cooldown), otherwise throw `InvalidArgument`
  - Check that the status stored in the program account is retracted
    otherwise throw `InvalidArgument`
  - In case a source program was provided (instruction account at index 2)
  which is not the program account:
    - Verify the source program account
    - Check that the status stored in the source program account is retracted,
    otherwise throw `InvalidArgument`
    - Check that the executable file stored in the source program account
    passes executable verification
      - The feature set that the executable file is verified against is not
      necessarily the current one, but the one of the epoch of the next slot
      - Also, during deployment certain deprecated syscalls are disabled,
      this stays the same as in the older loaders
    - Copy the entire source program account into the program account
    - Set the length of the source program account to zero
    - Transfer all funds of the  source program account to the program

>     account

Author: Thinking about it further: It is probably best to swap the funds of the source and the program account. That way the program account always has the right amount in the end, matching the size of the program and the rest is always at the source. Furthermore, the source should be assigned back to the system program so that the dApp devs can just transfer the rest of the funds out without any additional loader instructions.

>     account

Reviewer 8: Although I think leaving the extra funds in the source account has no problem, given that the developer still has the access to its key pair, I liked the idea of providing an extra account to reclaim the excessive SOL.

  - otherwise, if no source program was provided:
    - Check that the executable file stored in the program account passes
    executable verification
  - Change the slot in the program account to the current slot
  - Change the status stored in the program account to deployed

#### Retract

- Instruction accounts:
  - `[writable]` The program account to retract.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `4u32`
- Behavior:
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the slot stored in the program account is not the current
  (deployment cooldown), otherwise throw `InvalidArgument`
  - Check that the status stored in the program account is deployed,
    otherwise throw `InvalidArgument`
  - Note: The slot is **not** set to the current slot to allow a

>   retract-modify-redeploy-sequence within the same slot

Author: Yes, I will update the note.

  the header size of their loader (account owner) respectively

#### SetProgramLength

- Instruction accounts:
  - `[writable]` The program account to change the size of.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `2u32`
  - `u32` The new size after the operation.
- Behavior:
  - Charge 32 + new_size_in_bytes CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - If this is an initialization (program account length is too short to
  contain the header):
    - the owner of the program account is loader-v4,
    otherwise throw `InvalidAccountOwner`
    - the program account is writable, otherwise throw `InvalidArgument`
    - the provided authority (instruction account at index 1) signed,
    otherwise throw `MissingRequiredSignature`
  - If this is not an initialization:
    - Verify the program account
    - Check that the status stored in the program account is
    `NeverBeenDeployed` or `Retracted`,
    otherwise throw `InvalidArgument`
  - Check that there are enough funds in the program account for rent
  exemption of the new length, otherwise throw `InsufficientFunds`

>   - Set the length of the program account to the requested new size plus

Reviewer 9: Is it worth specifying special behavior in CPI to limit the new size to `size_at_start_of_top_level_instruction + 10kb`?

Reviewer 2: I don't think this was addressed

- Instruction data:
  - Enum variant `4u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Check that program account and recipient account do not alias,
    otherwise throw `AccountBorrowFailed`
  - Verify the program account
  - Check that the status stored in the program account is
  `NeverBeenDeployed` or `Retracted`,
  otherwise throw `InvalidArgument`
  - Set the length of the program account to 0 (removing the header too)
  - Transfer all lamports from the program account to the recipient account

#### Deploy

- Instruction accounts:
  - `[writable]` The program account to deploy.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `5u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the slot stored in the program account is not the current
  (deployment cooldown), otherwise throw `InvalidArgument`
  - Check that the status stored in the program account is `NeverBeenDeployed`
  or `Retracted`
    otherwise throw `InvalidArgument`

>   - Charge program_length_in_bytes CUs

Reviewer 9: Accounts can be 10MB, but the max CUs for a transaction is 1.4M CUs. So all costs that use the program length will need to be reduced

Author: Good point, should probably divide by `cpi_bytes_per_unit`

  - Check the status stored in the program account is `NeverBeenDeployed` or
  `Retracted`,
  otherwise throw `InvalidArgument`
  - Check that the source account is owned by loader v2, v3 or v4,
  otherwise throw `InvalidArgument`
  - and look-up the source header size:
    - loader-v2: 0 bytes
    - loader-v3 buffer: 37 bytes
    - loader-v3 programdata: 45 bytes
    - loader-v4: 48 bytes
  - Check that the source end offset (sum of source offset and length) does
  not exceed the maximum (source account length minus the source header size),
  otherwise throw `AccountDataTooSmall`
  - Check that the destination end offset (sum of destination offset and
  length) does not exceed the maximum (program account length minus the loader-v4
  header size), otherwise throw `AccountDataTooSmall`
  - Copy the chunk between the program accounts at the offsets, each shifted by
  the header size of their loader (account owner) respectively

#### SetProgramLength

- Instruction accounts:
  - `[writable]` The program account to change the size of.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `2u32`
  - `u32` The new size after the operation.
- Behavior:

>   - Charge 32 + new_size_in_bytes / cpi_bytes_per_unit CUs

Author: It is on purpose. The runtime has to memcpy all the existing data when reallocating a vector.

    otherwise throw `AccountBorrowFailed`
  - Verify the program account
  - Transfer lamports which are not needed for rent exemption from the
  program account to the recipient account

#### WithdrawAllLamports

- Instruction accounts:
  - `[writable]` The program account to withdraw from.
  - `[signer]` The authority of the program.
  - `[]` The recipient account.
- Instruction data:
  - Enum variant `4u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Check that program account and recipient account do not alias,
    otherwise throw `AccountBorrowFailed`
  - Verify the program account
  - Check that the status stored in the program account is
  `NeverBeenDeployed` or `Retracted`,
  otherwise throw `InvalidArgument`
  - Set the length of the program account to 0 (removing the header too)
  - Transfer all lamports from the program account to the recipient account

#### Deploy

- Instruction accounts:
  - `[writable]` The program account to deploy.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `5u32`
- Behavior:

>   - Charge 32 CUs

Author: Another CU charge happens later, a few lines down

otherwise throw `Immutable`

### Execution / Invocation

Invoking programs owned by loader-v4 checks in the following order that:

- the owner of the program account is loader-v4
- the program account is at least as long enough for the header
- the status stored in the program account is `Deployed` or `Finalized`
- the program account was not deployed within the current slot (delay
visibility)
- the executable file stored in the program account passes executable
verification

failing any of the above checks must throw `UnsupportedProgramId`.

### Program Management Instructions

#### Write

- Instruction accounts:
  - `[writable]` The program account to write to.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `0u32`
  - `u32` Byte offset at which to write the given bytes
  - `[u8]` Chunk of the programs executable file
- Behavior:
  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
  - Check there are at least two instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check the status stored in the program account is `NeverBeenDeployed` or

>   `Retracted`,

Reviewer 4: As per discussion offline (again), we're going to let it ride with the CLI. The goal is that this would unlock some efficient tooling, especially for power users like asm program devs.

  header size), otherwise throw `AccountDataTooSmall`
  - Copy the chunk between the program accounts at the offsets, each shifted by
  the header size of their loader (account owner) respectively

#### SetProgramLength

- Instruction accounts:
  - `[writable]` The program account to change the size of.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `2u32`
  - `u32` The new size after the operation.
- Behavior:
  - Charge 32 + new_size_in_bytes / cpi_bytes_per_unit CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account
  - Check that the status stored in the program account is
  `NeverBeenDeployed` or `Retracted`,
  otherwise throw `InvalidArgument`
  - Check that there are enough funds in the program account for rent
  exemption of the new length,
  otherwise throw `InsufficientFunds`
  - Set the length of the program account to the requested new size plus
  the header size

#### WithdrawExcessLamports

- Instruction accounts:
  - `[writable]` The program account to withdraw from.
  - `[signer]` The authority of the program.

>   - `[]` The recipient account.

Author: We can, but we don't have to require the recipient to be writable, right?

>   - `[]` The recipient account.

Reviewer 4: It has to be writable in order to be credited lamports.

>   - `[]` The recipient account.

Author: Yes you are right, sorry, confused it with the owner exception for crediting lamports.

explicates these steps in a Retract-Modify-Redeploy sequence of instructions
that allows multiple modifications to share one verification of the ELF in the
end. This sequence can be a single transaction or split across multiple
transactions.

Another issue with loader-v3 is that the executable file stored in the
programdata account is misaligned relative to the beginning of the account.
This currently requires a copy in the ELF loader to re-align the program.
To avoid any other alignment issues all states of the accounts owned by a
loader should have the same layout and only be differentiated by a status enum.

Additionally, there currently is no complete specification of the loaders
program management instructions. This proposal would thus fill that gap once
loader-v4 goes into production.

See impact for further motivation.

## Alternatives Considered

### Hot swap redeployment

A delay-visibility-free redeployment could be achieved by keeping the swap
program around until the end of the slot. This would however mean that two
accounts per program must be loaded until the dapp developer reclaims the
second one. That would defy the purpose of this proposal which is to get rid

> of the proxy account.

Author: While the delay visibility restriction is implemented in the loaders, it owns its existence to the way we handle the index structure of the global program cache. Thus, we have to fix it there first and then could later lift the restriction in the loaders. Kind of unrelated to this SIMD though.

> of the proxy account.

Reviewer 2: Ah yeah, makes sense 👍🏻

its file is, it has to be parsed and calculated from offsets of various
sections, which increases the validator implementation complexity. Second,
this would require the ELF header to always be uploaded first and always in a
single chunk, wich increases the complexity of the uploading logic. Instead we
choose the user having to announce the length of the program to be uploaded
explicitly before the actual upload starts.

## New Terminology

The _current slot_ is as in the Clock sysvar.

## Detailed Design

The feature gate must enable loader-v4 program management and execution.

### Owned Program Accounts

Accounts of programs owned by loader-v4 must have the following layout:

- Header (which is 48 bytes long):
  - `u64` status enum:
    - Enum variant `0u64`: `Invalid`, account was zero-filled externally
    - Enum variant `1u64`: `NeverBeenDeployed`, used as write buffer
    - Enum variant `2u64`: `Retracted`, program is in maintenance
    - Enum variant `3u64`: `Deployed`, program is ready to be executed
    - Enum variant `4u64`: `Finalized`, same as `Deployed`, but can not be
    modified anymore
  - `u64` Slot in which the program was last deployed, retracted or

>   initialized.

Reviewer 2: This doesn't look accurate anymore. It should only represent the last deployment slot right?

  - `[u8; 32]` Authority address which can send program management
  instructions.
- Body:
  - `[u8]` The programs executable file

Verification the program account checks in the following order that:

- the owner of the program account is loader-v4,
otherwise throw `InvalidAccountOwner`
- the program account is at least as long enough for the header,
otherwise throw `AccountDataTooSmall`
- the program account is writable, otherwise throw `InvalidArgument`
- the provided authority (instruction account at index 1) signed,
otherwise throw `MissingRequiredSignature`
- the authority stored in the program account is the one provided,
otherwise throw `IncorrectAuthority`
- the status stored in the program account is not `Invalid`,
otherwise throw `InvalidArgument`
- the status stored in the program account is not `Finalized`,

> otherwise throw `Immutable`

Reviewer 2: Should be ok to withdraw excess lamports from finalized programs right? If we wanted to allow this, we could keep the authority as-is without reseting to the system program.

  not exceed the maximum (program account length minus the header size),
  otherwise throw `AccountDataTooSmall`
  - Copy the chunk into the program account at the offset shifted by the
  header size

#### Copy

- Instruction accounts:
  - `[writable]` The program account to copy to.
  - `[signer]` The authority of the program.
  - `[]` The account to copy from.
- Instruction data:
  - Enum variant `1u32`
  - `u32` Byte offset at which to write
  - `u32` Byte offset at which to read
  - `u32` Length of the chunk to copy in bytes
- Behavior:
  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
  - Check there are at least three instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Check that program account and source account do not alias,
  otherwise throw `AccountBorrowFailed`
  - Verify the program account
  - Check the status stored in the program account is `NeverBeenDeployed` or
  `Retracted`,
  otherwise throw `InvalidArgument`
  - Check that the source account is owned by loader v2, v3 or v4,
  otherwise throw `InvalidArgument`
  - and look-up the source header size:
    - loader-v2: 0 bytes
    - loader-v3 buffer: 37 bytes

>     - loader-v3 programdata: 45 bytes

Reviewer 2: Can you also specify that a non-buffer / non-programdata loader-v3 owned account will be rejected with `InvalidArgument`? (same for uninitialized loader-v4 account)

- Instruction accounts:
  - `[writable]` The program account to withdraw from.
  - `[signer]` The authority of the program.
  - `[writable]` The recipient account.
- Instruction data:
  - Enum variant `4u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Check that program account and recipient account do not alias,
    otherwise throw `AccountBorrowFailed`
  - Check that the recipient account is writable,
  otherwise throw `InvalidArgument`
  - Verify the program account
  - Check that the status stored in the program account is
  `NeverBeenDeployed` or `Retracted`,
  otherwise throw `InvalidArgument`
  - Set the length of the program account to 0 (removing the header too)
  - Transfer all lamports from the program account to the recipient account

#### Deploy

- Instruction accounts:
  - `[writable]` The program account to deploy.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `5u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account

>   - Check that the slot stored in the program account is not the current

Reviewer 2: We only need this check if the program is `Retracted`. If the current slot is zero (in block 0) and the status is `NeverBeenDeployed` then the deploy should still be allowed.

>   (deployment cooldown), otherwise throw `InvalidArgument`

Reviewer 2: Just curious, is this even possible? A program can't be reclaimed if it was deployed in the current slot, right?

Reviewer 2: Not sure I'm following. Yes, we have a check in `Retract` that the last deployed slot is not the same as the current slot. So if you cannot retract in the same slot that a program was deployed in, then it's impossible during `Deploy` that the last deployment slot is the current slot. So we don't need to check the last deployment slot during `Deploy` technically.

otherwise throw `AccountDataTooSmall`
- the program account is writable, otherwise throw `InvalidArgument`
- the provided authority (instruction account at index 1) signed,
otherwise throw `MissingRequiredSignature`
- the authority stored in the program account is the one provided,
otherwise throw `IncorrectAuthority`
- the status stored in the program account is not `Uninitalized`,
otherwise throw `InvalidArgument`
- the status stored in the program account is not `Finalized`,
otherwise throw `Immutable`

### Execution / Invocation

Invoking programs owned by loader-v4 checks in the following order that:

- the owner of the program account is loader-v4
- the program account is at least as long enough for the header
- the status stored in the program account is `Deployed` or `Finalized`
- the program account was not deployed within the current slot (delay
visibility)
- the executable file stored in the program account passes executable
verification. The executable file starts after the 48 byte header and has the
length defined in the header, which could be less than the account length.

failing any of the above checks must throw `UnsupportedProgramId`.

> ### Program Management Instructions

Reviewer 2: Should also add a note saying that instruction data length is limited to 1232 bytes to match loader-v3 behavior

> The loader-v4 intructions Deploy and Retract are not authorized in CPI.

Author: Was thinking about the CPI restriction some more: Isn't it a problem for multisig and governance? E.g. https://github.com/solana-foundation/solana-improvement-documents/pull/167#discussion_r2409321450 would not work, right?

> The loader-v4 intructions Deploy and Retract are not authorized in CPI.

Reviewer 2: Ah yeah, you're right. Didn't think about that. Let's try to remove this restriction then. If I remember correctly it's because we don't yet allow changing the executable flag from CPI.

  - Check there are at least three instruction accounts,
  otherwise throw `NotEnoughAccountKeys`
  - Check that program account and source account do not alias,
  otherwise throw `AccountBorrowFailed`
  - Verify the program account header
  - Check the status stored in the program account is `NeverBeenDeployed` or
  `Retracted`,
  otherwise throw `InvalidArgument`
  - Check that the source accounts owner and look-up the source header size:
    - loader-v2: 0 bytes
    - loader-v3 buffer: 37 bytes
    - loader-v3 programdata: 45 bytes
    - loader-v4: 48 bytes
    - if none of the above matches throw `InvalidArgument`
  - Check that the source end offset (sum of source offset and length) does
  not exceed the maximum (source account length minus the source header size),
  otherwise throw `AccountDataTooSmall`
  - Check that the destination end offset (sum of destination offset and
  length) does not exceed the maximum (program account length minus the loader-v4
  header size), otherwise throw `AccountDataTooSmall`
  - Copy the chunk between the program accounts at the offsets, each shifted by
  the header size of their loader (account owner) respectively

> #### SetProgramLength

Reviewer 2: If we stored the length of the program data in the program header, this instruction could be run on programs without retracting first. This would make program upgrades a little easier for smart contract controlled wallets because the instruction could be permissionless (as is the case in loader v3) and therefore not need to be done via CPI. I think this would be a big ergonomics improvement

Reviewer 2: The primary benefit is that it can be called without CPI if the upgrade authority is a PDA

    otherwise throw `InvalidArgument`
  - Change the status stored in the program account to `Finalized`

#### SetAccountLength

- Instruction accounts:
  - `[writable]` The program account to change the size of.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `3u32`
  - `u32` The new size after the operation.
- Behavior:

>   - Charge 32 + new_size_in_bytes / cpi_bytes_per_unit CUs

Author: I think we discussed this before. The reason why we charge for the entire length is because the program runtime has to copy over the existing data into the new allocation.

>   - Charge 32 + new_size_in_bytes / cpi_bytes_per_unit CUs

Reviewer 2: Ah yeah, you're right never mind.

  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account header,
  but skip the signer check if the new size is greater than the current size
  - Check that the status stored in the program account is
  `NeverBeenDeployed` or `Retracted`,
  otherwise throw `InvalidArgument`
  - Check that the new size in bytes is at least the header length (48 bytes),
  otherwise throw `InvalidArgument`
  - Set the length of the program account to the requested new size.
  - Note: In CPI the maximum growth is limited to 10 KiB in ABI v1 and
  0 bytes in ABI v0. Thus, this should be called from a top-level instruction.
  For this purpose this operation does not check the signer when growing.

#### Write

- Instruction accounts:
  - `[writable]` The program account to write to.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `4u32`
  - `u32` Byte offset at which to write the given bytes

>   - `[u8]` Chunk of the programs executable file

Reviewer 2: This looks different from loader-v3 which serializes the length of the chunk as a `u64` before the chunk data. Are you implying that the length isn't serialized here and that `chunk_length_in_bytes` is inferred from `ix.data.len - 4 bytes (enum variant) - 4 bytes (offset)`? I think we should match loader-v3 behavior and serialize the chunk length as well. But consider serializing the length as u16 or u32 instead of u64 to save some bytes

Author: What is the advantage of doing so? It would introduce additional failure cases of the explicit field and implicit calculation disagreeing.

  - Check that the source end offset (sum of source offset and length) does
  not exceed the maximum (source account length minus the source header size),
  otherwise throw `AccountDataTooSmall`
  - Check that the destination end offset (sum of destination offset and
  length) does not exceed the maximum (program account length minus the loader-v4
  header size), otherwise throw `AccountDataTooSmall`
  - Copy the chunk between the program accounts at the offsets, each shifted by
  the header size of their loader (account owner) respectively

#### Deploy

- Instruction accounts:
  - `[writable]` The program account to deploy.
  - `[signer]` The authority of the program.
- Instruction data:
  - Enum variant `6u32`
  - `u32` The executable length in bytes.
- Behavior:
  - Charge 32 CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account header
  - Check that the status stored in the program account is `NeverBeenDeployed`
  or `Retracted`
    otherwise throw `InvalidArgument`
  - Check that the executable length is not longer than the program account,
    otherwise throw `InvalidArgument`

>   - Charge program_length_in_bytes / cpi_bytes_per_unit CUs

Reviewer 2: We should only charge for the executable length rather than the full account data length

```suggestion
  - Charge executable length / cpi_bytes_per_unit CUs
```

  - Check that the executable file stored in the program account passes
  executable verification

>   - Change the executable length in the program account header

Reviewer 2: ```suggestion
  - Change the executable length in the program account header to the specified
    value.
```

  retract-modify-redeploy-sequence within the same slot or even within the
  same transaction.
  - Change the status stored in the program account to `Retracted`
  - Set the `is_executable` flag to `false`

#### Initialize

- Instruction accounts:
  - `[writable]` The account to initialize as program account.
  - `[signer]` The new authority of the program.
- Instruction data:
  - Enum variant `7u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Check that the owner of the program account is loader-v4,
    otherwise throw `InvalidAccountOwner`
  - Check that the program account is writable,
    otherwise throw `InvalidArgument`
  - Check that there are enough funds in the program account for rent
  exemption of the header,
    otherwise throw `InsufficientFunds`
  - Check that the new authority (instruction account at index 1) has signed,
    otherwise throw `MissingRequiredSignature`
  - Set the length of the program account to the header size

>   - Set the slot to zero, **not** the current slot

Reviewer 2: Why is this important? I don't think it matters what the slot is set to for accounts with `NeverBeenDeployed` status right?

  - `[signer]` The current authority of the program.
  - `[signer]` The new authority of the program.
- Instruction data:
  - Enum variant `8u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least three instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account header
  - Check that the new authority (instruction account at index 2) has signed,
    otherwise throw `MissingRequiredSignature`
  - Check that the current authority is different from the new authority,
    otherwise throw `InvalidArgument`
  - Copy the new authority address into the program account

#### Finalize

- Instruction accounts:
  - `[writable]` The program account to finalize.
  - `[signer]` The current authority of the program.
- Instruction data:
  - Enum variant `9u32`
- Behavior:
  - Charge 32 CUs
  - Check there are at least two instruction accounts,
    otherwise throw `NotEnoughAccountKeys`
  - Verify the program account header
  - Check that the status stored in the program account is `Deployed` or
    that the status is `Retracted` and the program length is 0 (header only),
    otherwise throw `InvalidArgument`

>   - Change the status stored in the program account to `Finalized`

Reviewer 2: What do you think about adding an additional `RetractedFinalized` status too? The current design of `Finalize` makes a finalized `Retracted` program look kinda executable from looking at the status alone. It requires the runtime to verify the program (ie check the program length is 0) to know that it's actually retracted.

Author: > It requires the runtime to verify the program (ie check the program length is 0) to know that it's actually retracted.

That is only once in the `Finalize` instruction and even if we added an additional enum item this would remain the case because the `Finalize` instruction is what sets that status. `Finalized` and `RetractedFinalized` would behave exactly the same in all loader instructions and you always have to load the account data anyway to see the status, thus you always know the length too. I don't really see the benefit of distinguishing between these two specifically yet.

Reviewer 2: Yeah fair point about needing to load the account data to see the status anyways. I'm fine with keeping it as a single `Finalized` status and distinguishing "retracted" by checking program data length is zero
